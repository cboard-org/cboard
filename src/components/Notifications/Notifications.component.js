import React from 'react';
import PropTypes from 'prop-types';
import Snackbar from '@material-ui/core/Snackbar';
import Alert from '@material-ui/lab/Alert';
import Button from '@material-ui/core/Button';
import { NOTIFICATION_DELAY } from './Notifications.constants';
import { FormattedMessage } from 'react-intl';
import messages from './Notifications.messages';
import { Link } from 'react-router-dom';

const propTypes = {
  config: PropTypes.object.isRequired,
  handleNotificationDismissal: PropTypes.func.isRequired,
  message: PropTypes.string.isRequired,
  open: PropTypes.bool.isRequired,
  kind: PropTypes.string
};

function onRefreshPage() {
  window.location.reload(true);
}

const Notifications = ({
  config,
  handleNotificationDismissal,
  message,
  open,
  kind
}) => {
  const childrenDependKind = kind => {
    if (!kind) return null;
    if (kind === 'refresh')
      return (
        <Alert
          elevation={6}
          variant="filled"
          onClose={handleNotificationDismissal}
          severity="info"
          action={
            <Button variant="outlined" onClick={onRefreshPage}>
              <FormattedMessage {...messages.refreshPage} />
            </Button>
          }
        >
          <span id="message-id">{message}</span>
        </Alert>
      );
    if (kind === 'cloudSpeakError' || kind === 'cloudVoiceIsSeted') {
      const isCloudError = !!(kind === 'cloudSpeakError');
      return (
        <Alert
          variant="filled"
          severity={isCloudError ? 'error' : 'info'}
          style={{
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center'
          }}
          action={
            isCloudError ? (
              <Button
                size="small"
                variant="outlined"
                style={{
                  color: 'white',
                  borderColor: 'white',
                  textAlign: 'center'
                }}
                component={Link}
                to="/settings/speech"
              >
                <FormattedMessage {...messages.changeVoiceOnError} />
              </Button>
            ) : null
          }
        >
          {isCloudError ? (
            <FormattedMessage {...messages.cloudSpeakErrorAlert} />
          ) : (
            <FormattedMessage {...messages.cloudVoiceIsSetedAlert} />
          )}
        </Alert>
      );
    }
  };
  return (
    <Snackbar
      {...config}
      open={open}
      ContentProps={{
        variant: 'elevation',
        'aria-describedby': 'message-id'
      }}
      message={<span id="message-id">{message}</span>}
      autoHideDuration={NOTIFICATION_DELAY}
      onClose={handleNotificationDismissal}
    >
      {childrenDependKind(kind)}
    </Snackbar>
  );
};

Notifications.propTypes = propTypes;

export default Notifications;
