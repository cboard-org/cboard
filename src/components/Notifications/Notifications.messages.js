import { defineMessages } from 'react-intl';

export default defineMessages({
  refreshPage: {
    id: 'cboard.components.Notifications.refreshPage',
    defaultMessage: 'Refresh page'
  },
  cloudSpeakErrorAlert: {
    id: 'cboard.components.Notifications.cloudSpeakErrorAlert',
    defaultMessage:
      'Your internet connection is insufficient to reproduce an online voice properly.'
  },
  changeVoiceOnError: {
    id: 'cboard.components.Notifications.changeVoiceOnError',
    defaultMessage: 'Change the voice'
  },
  cloudVoiceIsSetedAlert: {
    id: 'cboard.components.Notifications.cloudVoiceIsSetedAlert',
    defaultMessage:
      'An online voice was set. An Internet connection is required during its use.'
  }
});
