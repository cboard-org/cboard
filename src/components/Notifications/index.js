export { default } from './Notifications.container';
