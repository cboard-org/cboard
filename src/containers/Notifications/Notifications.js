import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Snackbar from 'material-ui/Snackbar';
import {
  hideNotification as hideNotificationActionCreator,
  showNotification as showNotificationActionCreator
} from './actions';

class Notifications extends Component {
  // maintain queued notifications
  // a notification is queued if already another
  // notification is already displayed
  // this is not part of the state beacuase
  // we do not want a change in this to re-render
  // our component
  queuedNotifications = [];

  shouldComponentUpdate(nextProps) {
    // following the material design guidelines
    // we can only show one notification at a time
    // here we ignore updates to an already displayed
    // notifcation and instead queue them to be shown
    // after the present one transitions out
    if (
      this.props.open &&
      nextProps.open &&
      this.props.message !== nextProps.message
    ) {
      this.queuedNotifications.push(nextProps.message);
      return false;
    } else {
      return true;
    }
  }

  handleNotificationDismissal = (event, reason) => {
    this.props.hideNotification();
  };

  showQueuedNotificationIfAny = () => {
    if (this.queuedNotifications.length !== 0) {
      console.log('before', this.queuedNotifications);
      this.props.showNotification(this.queuedNotifications[0]);
      this.queuedNotifications.splice(0, 1);
      console.log('after', this.queuedNotifications);
    }
  };

  render() {
    const {
      message,
      open,
      showNotification,
      hideNotification,
      ...config
    } = this.props;

    if (message && message.length < 1) {
      return null;
    }

    return (
      <Snackbar
        {...config}
        open={open}
        message={message}
        onRequestClose={this.handleNotificationDismissal}
        // show any queued notifications after the
        // present one transitions out
        onExited={this.showQueuedNotificationIfAny}
      />
    );
  }
}

Notifications.defaultProps = {
  anchorOrigin: {
    vertical: 'bottom',
    horizontal: 'right'
  },
  autoHideDuration: 3000, // ms
  open: false
};

Notifications.propTypes = {
  message: PropTypes.string.isRequired,
  open: PropTypes.bool.isRequired,
  showNotification: PropTypes.func.isRequired,
  hideNotification: PropTypes.func.isRequired
};

function mapStateToProps({ notification: { message, open } }) {
  return {
    message,
    open
  };
}

function mapDispatchToProps(dispatch) {
  return {
    showNotification(message) {
      dispatch(showNotificationActionCreator(message));
    },

    hideNotification() {
      dispatch(hideNotificationActionCreator());
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Notifications);
