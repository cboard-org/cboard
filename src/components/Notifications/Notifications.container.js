import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { hideNotification, showNotification } from './Notifications.actions';
import Notifications from './Notifications.component';

class NotificationsContainer extends Component {
  static propTypes = {
    /**
     * If true, notification bar is open, used by showNotification
     */
    open: PropTypes.bool.isRequired,
    /**
     * The Message to display
     */
    message: PropTypes.string.isRequired,
    /**
     * Shows notification bar
     */
    showNotification: PropTypes.func.isRequired,
    /**
     * Hids notification bar
     */
    hideNotification: PropTypes.func.isRequired
  };

  static defaultProps = {
    anchorOrigin: {
      vertical: 'bottom',
      horizontal: 'center'
    },
    autoHideDuration: 5000, // ms
    open: false
  };

  // maintain queued notifications
  // a notification is queued if already another
  // notification is already displayed
  // this is not part of the state beacuase
  // we do not want a change in this to re-render
  // our component
  queuedNotifications = [];

  shouldComponentUpdate(nextProps) {
    // following the material design guidelines
    // we can only show one notification at a time
    // here we ignore updates to an already displayed
    // notification and instead queue them to be shown
    // after the present one transitions out
    if (
      this.props.open &&
      nextProps.open &&
      this.props.message !== nextProps.message
    ) {
      this.queuedNotifications.push(nextProps.message);
      return false;
    } else {
      return true;
    }
  }

  handleNotificationDismissal = (event, reason) => {
    const { hideNotification } = this.props;
    if (reason === 'clickaway') {
      return;
    }
    hideNotification();
  };

  showQueuedNotificationIfAny = () => {
    if (this.queuedNotifications.length !== 0) {
      console.log('before', this.queuedNotifications);
      this.props.showNotification(this.queuedNotifications[0]);
      this.queuedNotifications.splice(0, 1);
      console.log('after', this.queuedNotifications);
    }
  };

  render() {
    const {
      open,
      message,
      showNotification,
      hideNotification,
      ...config
    } = this.props;

    if (message && message.length < 1) {
      return null;
    }

    return (
      <Notifications
        config={config}
        open={open}
        message={message}
        handleNotificationDismissal={this.handleNotificationDismissal}
        showQueuedNotificationIfAny={this.showQueuedNotificationIfAny}
      />
    );
  }
}

const mapStateToProps = ({ notification: { message, open } }) => ({
  message,
  open
});

const mapDispatchToProps = {
  showNotification,
  hideNotification
};

export default connect(mapStateToProps, mapDispatchToProps)(
  NotificationsContainer
);
