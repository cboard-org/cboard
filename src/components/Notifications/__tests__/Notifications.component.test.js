import React from 'react';
import { shallowMatchSnapshot } from '../../../common/test_utils';
import Notifications from '../Notifications.component';

const COMPONENT_PROPS = {
  config: {},
  handleNotificationDismissal: () => {},
  message: 'hhh',
  open: true,
  showQueuedNotificationIfAny: () => {}
};

describe('Notifications tests', () => {
  test('default renderer', () => {
    shallowMatchSnapshot(<Notifications {...COMPONENT_PROPS} />);
  });
});
