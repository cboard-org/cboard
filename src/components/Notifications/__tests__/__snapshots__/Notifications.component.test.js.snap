// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications tests default renderer 1`] = `
<WithStyles(Snackbar)
  ContentProps={
    Object {
      "aria-describedby": "message-id",
      "variant": "body1",
    }
  }
  autoHideDuration={5000}
  message={
    <span
      id="message-id"
    >
      hhh
    </span>
  }
  onClose={[Function]}
  onExited={[Function]}
  open={true}
/>
`;
